src/policy/ancestor_packages.h

```diff
@@ -0,0 +1,167 @@
+     * provided through LinearizeWithFees(), using mining_sequence scores. */
+    std::vector<std::reference_wrapper<PackageEntry>> m_txns;
+
+    /** Helper function for recursively constructing ancestor caches in ctor. */
+    void visit(const CTransactionRef&);
+public:
+    /** Constructs ancestor package, sorting the transactions topologically and constructing the
+     * txid_to_tx and ancestor_subsets maps. It is ok if the input txns is not sorted.
+     * Expects:
+     * - No duplicate transactions.
+     * - No conflicts between transactions.
+     */
+    AncestorPackage(const Package& txns_in);
+
+    bool IsAncestorPackage() const { return m_ancestor_package_shaped; }
+
+    /** Returns all of the transactions, linearized. */
+    Package Txns() const;
+
+    /** Returns all of the transactions, without the skipped and dangling ones, linearized. */
+    Package FilteredTxns() const;
+
+    /** Get the sorted, filtered ancestor subpackage for a tx. Includes the tx. Does not include
+     * ancestors that are MEMPOOL. If this tx or any ancestors are DANGLING, returns std::nullopt. */
+    std::optional<std::vector<CTransactionRef>> FilteredAncestorSet(const CTransactionRef& tx) const;
+
+    /** Get the total fee and vsize of the ancestor subpackage for a tx. Includes the tx. Does not
+     * include ancestors that are MEMPOOL. If this tx or any ancestors are DANGLING, returns
+     * std::nullopt. This result is always consistent with FilteredAncestorSet(). */
+    std::optional<std::pair<CAmount, int64_t>> FilteredAncestorFeeAndVsize(const CTransactionRef& tx) const;
+
+    /** Get the fee and vsize of a tx. Returns std::nullopt if this information is unknown or this
+     * tx is DANGLING. */
+    std::optional<std::pair<CAmount, int64_t>> GetFeeAndVsize(const CTransactionRef& tx) const;
+
+    /** Should be called if a tx or same txid transaction is found in mempool or submitted to it.
+     * From now on, skip this tx from any result in FilteredAncestorSet(). Does not affect Txns().
+     */
+    void IsInMempool(const CTransactionRef& transaction);
+
+    /** Should be called if a tx will not be considered because it is missing inputs or is
+     * conflicting with another transaction that is already in mempool.
+     * Skip a transaction and all of its descendants. From now on, if this transaction is present
+     * in the ancestor set, FilteredAncestorSet() returns std::nullopt for that tx. Does not affect Txns(). */
+    void IsDanglingWithDescendants(const CTransactionRef& transaction);
+
+    /** Add information about fee and vsize for a transaction. */
+    void AddFeeAndVsize(const uint256& txid, CAmount fee, int64_t vsize);
+
+    /** Re-linearize transactions using the fee and vsize information given. Updates Txns().
+     * Information must have been provided for all non-skipped transactions via AddFeeAndVsize().
+     * @returns true if successful, false if something went wrong. */
```

Comment:
Would be helpful to explain what can go wrong? What are you meant to do if you get a `false` result? Punish the peer for giving you bad info? Do some sort of orphan resolution? assert because you've run out of memory or ended up with corrupt data structures?

Reply:
Added doc.